base/error_response_transport.go

```diff
@@ -0,0 +1,128 @@
+/**
+ * © Copyright IBM Corporation 2024. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package base
+
+import (
+	"bytes"
+	"encoding/json"
+	"io"
+	"net/http"
+	"strings"
+)
+
+// ErrorResponse is an error augmentation RoundTripper
+// that in case of http error response parses the return body
+// and converts it to a closer match to the standard error including adding a trace ID
+// and appending the CouchDB/Cloudant error reason to the message
+type ErrorResponse struct {
+	next http.RoundTripper
+}
+
+// NewErrorResponse creates a new ErrorResponse middleware
+func NewErrorResponse(rt http.RoundTripper) http.RoundTripper {
+	return &ErrorResponse{next: rt}
+}
+
+// RoundTrip implements RoundTripper interface
+func (er *ErrorResponse) RoundTrip(req *http.Request) (resp *http.Response, err error) {
+	defer func() {
+		if err == nil && resp.StatusCode >= http.StatusBadRequest {
+			ct := resp.Header.Get("Content-Type")
+			if strings.HasPrefix(ct, "application/json") {
+				err = transformError(resp)
+				if err != nil {
+					resp = nil
+				}
+			}
+		}
+	}()
+
+	return er.next.RoundTrip(req)
+}
+
+// drainBody reads body into memory and returns two identical ReadClosers
+// one of which is safe to read
+func drainBody(body io.ReadCloser) (orig, save io.ReadCloser, err error) {
+	if body == nil || body == http.NoBody {
+		return http.NoBody, http.NoBody, nil
+	}
+	var buf bytes.Buffer
+	if _, err = buf.ReadFrom(body); err != nil {
+		return nil, body, err
+	}
+	if err = body.Close(); err != nil {
+		return nil, body, err
+	}
+	r := bytes.NewReader(buf.Bytes())
+	return io.NopCloser(&buf), io.NopCloser(r), nil
+}
+
+// transformError reads the response's body, parses it as json, augments it,
+// encodes back and sets as a new response
+func transformError(resp *http.Response) error {
+	var err error
+	var save io.ReadCloser
+	savecl := resp.ContentLength
```

Comment:
This makes me a little nervous for 2 reasons:
1. streaming cases don't have a content-length
2. compressed vs uncompressed

My reading is that this is a save of the pre-transformation content length and that we just use the same value if we don't transform. So for case 1 (no length) with no transform  I think that's fine assuming we aren't changing the body to some type that can't be streamed (which I don't think is the case). For case 1 where we do transform we are adjusting the length fo the new body regardless of whether it was previously set or not, which sort of changes a streaming case to a non-streaming case, but I think it's largely irrelevant because the body is already read by this point. Is that right?

For case 2, I think it matters if transparent decompression is completed by this point (which I hope that it is - otherwise the body reading is going to need more code)! Also it might matter if the transparent decompression updates the content-length to the uncompressed size before we see it if we subsequently set it on an untransofmred response. That is to say if the compressed content-length is `15` and that is the value we save, but we write a saved decompressed body with length `24` then we'd have a mismatch.

I expect I'm worrying about nothing, but it would be good to know if you've validated the behaviour.

Reply:
For 1 yes, we are setting `content-length` to whatever it was before and if we are changing body we are setting real length, but it is irrelevant because 1) we already read body which is an error anyway and not _really_ streamed 2) go's `AsStream` is kind of a lie, it's the same `ReaderCloser` on body in any of cases, `AsStream` just returns it instead of reading and deserializing and we are returning honest `ReaderCloser` just with our own  error, i.e. we are not altering API behaviour. I guess I can add detection for `Transfer-Encoding` set to chunked and skip setting `content-lenght`, just because. I doubt it'd make a real difference.

For 2, we are doing our manipulation in `defer` and `Transport` automatically handles decompression, so we should be good. It is possible to turn off this automatic decompression by setting `Accept-Encoding`, but in this case we'll not be able to deserialize response and just proxy it through.

Reply:
OK sounds like this is fine as is.